x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -278,15 +285,12 @@ public DatafeedConfig(StreamInput in) throws IOException {
         this.jobId = in.readString();
         this.queryDelay = in.readOptionalTimeValue();
         this.frequency = in.readOptionalTimeValue();
-        if (in.readBoolean()) {
-            this.indices = in.readCollectionAsImmutableList(StreamInput::readString);
-        } else {
-            this.indices = null;
-        }
+        this.indices = in.readOptionalStringCollectionAsList();
         // each of these writables are version aware
-        this.queryProvider = QueryProvider.fromStream(in);
+        this.queryProvider = in.readOptionalWriteable(QueryProvider::fromStream);
```

Comment:
If I understand these [failing tests](https://buildkite.com/elastic/elasticsearch-pull-request/builds/10270#018d8db2-efd6-451c-b160-f70e1e412108) correctly, this is breaking bwc (old/new clients cannot deserialize each other's datafeed config). Please correct me if I'm wrong.

What backward compatibility do we actually need? Is there something documented about that?

How do we fix such isuses? The query used to be mandatory, but now it isn't anymore, because you can either have a query or an esql query.

Reply:
Thanks for the explanation; that makes sense.

How do I ensure that every node in the cluster is as new as `TransportVersions.ESQL_IN_DATAFEEDS`? And what do I do if that's not the case? Let the create datafeed endpoint throw an exception with "not supported, because some nodes are old"? 

And what if (hypothetically) a node with an old version joins the cluster, after an ES|QL datafeed is defined and accepted?

Reply:
> How do I ensure that every node in the cluster is as new as `TransportVersions.ESQL_IN_DATAFEEDS`? And what do I do if that's not the case? Let the create datafeed endpoint throw an exception with "not supported, because some nodes are old"?

You can use `ClusterState.getMinTransportVersion()`. There's an example of something very similar here:

https://github.com/elastic/elasticsearch/blob/1f0ea3e015a789bdb71dc4c65b7bbf8f1bdf7d85/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java#L192-L201

The place that is done also shows a good place to do this. The check can be done in `TransportPutDatafeedAction` and `TransportUpdateDatafeedAction` where the cluster state will be available more easily than in the `Builder` for the config itself.

> And what if (hypothetically) a node with an old version joins the cluster, after an ES|QL datafeed is defined and accepted?

That node will be rejected. The master node will not allow a node to join the cluster if it has a version older than the nodes currently in the cluster.